File: lc_soln/0004_median_of_two_sorted_arrays/solution.py
from typing import List


class Solution:
    def findMedianSortedArrays(self, nums1: List[int], nums2: List[int]) -> float:
        n1 = len(nums1)
        n2 = len(nums2)
        if n1 > n2:
            nums1, nums2 = nums2, nums1
            n1, n2 = n2, n1

        half = (n1 + n2) // 2

        l = 0
        r = n1 - 1
        while True:
            mid = l + (r - l) // 2
            j = half - mid - 2

            left1 = nums1[mid] if mid >= 0 else float("-inf")
            right1 = nums1[mid + 1] if mid < n1 - 1 else float("inf")
            left2 = nums2[j] if j >= 0 else float("-inf")
            right2 = nums2[j + 1] if j < n2 - 1 else float("inf")

            if left1 <= right2 and left2 <= right1:
                if (n1 + n2) % 2 == 0:
                    return (max(left1, left2) + min(right1, right2)) / 2
                return min(right1, right2)
            elif left1 > right2:
                r = mid - 1
            else:
                l = mid + 1


def main(): ...


if __name__ == "__main__":
    main()

File: lc_soln/0004_median_of_two_sorted_arrays/solution.rs
struct Solution;

impl Solution {
    pub fn find_median_sorted_arrays(mut nums1: Vec<i32>, mut nums2: Vec<i32>) -> f64 {
        let mut n1 = nums1.len() as i32;
        let mut n2 = nums2.len() as i32;
        if n1 > n2 {
            (n1, n2) = (n2, n1);
            (nums1, nums2) = (nums2, nums1);
        }

        let half = (n1 + n2 + 1) / 2;

        let mut l = 0i32;
        let mut r = n1;
        loop {
            let i = l + (r - l) / 2;
            let j = half - i;

            let left1 = if i > 0 {
                nums1[(i - 1) as usize]
            } else {
                i32::MIN
            };
            let right1 = if i < n1 { nums1[i as usize] } else { i32::MAX };
            let left2 = if j > 0 {
                nums2[(j - 1) as usize]
            } else {
                i32::MIN
            };
            let right2 = if j < n2 { nums2[j as usize] } else { i32::MAX };

            if left1 <= right2 && left2 <= right1 {
                if (n1 + n2) % 2 == 0 {
                    return (left1.max(left2) + right1.min(right2)) as f64 / 2.0;
                }
                return left1.max(left2) as f64;
            } else if left1 > right2 {
                r = i - 1;
            } else {
                l = i + 1;
            }
        }
    }
}

fn main() {}
